File: sdks/typescript/src/types/query.ts
export interface QueryBody {
  query: string;
  top_k?: number;
  filters?: Record<string, unknown>;
  min_score?: number;
  search_mode?: "semantic" | "keyword" | "hybrid";
  rerank?: boolean;
}

export interface QueryResult {
  id: string;
  text: string;
  score: number;
  document_id: string | null;
  document_filename: string | null;
  chunk_index: number | null;
  page_no?: number | null;
  char_start?: number | null;
  char_end?: number | null;
  metadata: Record<string, unknown>;
}

export interface QueryTimings {
  embed_ms?: number;
  search_ms?: number;
  rerank_ms?: number;
  cache_ms?: number;
  total_ms?: number;
  cache_hit?: boolean;
}

export interface QueryResponse {
  results: QueryResult[];
  query: string;
  collection: string;
  total: number;
  timings?: QueryTimings;
}

export interface MultiQueryBody {
  query: string;
  collections: string[];
  top_k?: number;
  filters?: Record<string, unknown>;
  min_score?: number;
  search_mode?: "semantic" | "keyword" | "hybrid";
  rerank?: boolean;
}

export interface MultiQueryResult {
  id: string;
  text: string;
  score: number;
  document_id: string | null;
  document_filename: string | null;
  chunk_index: number | null;
  collection: string;
  metadata: Record<string, unknown>;
}

export interface MultiQueryResponse {
  results: MultiQueryResult[];
  query: string;
  collections: string[];
  total: number;
}

export interface BatchQueryItem {
  collection: string;
  query: string;
  top_k?: number;
  filters?: Record<string, unknown>;
  min_score?: number;
  search_mode?: "semantic" | "keyword" | "hybrid";
  rerank?: boolean;
}

export interface BatchQueryBody {
  queries: BatchQueryItem[];
}

export interface BatchQueryResultItem {
  results: QueryResult[];
  query: string;
  collection: string;
  total: number;
}

export interface BatchQueryResponse {
  results: BatchQueryResultItem[];
}

File: sdks/python/src/bigrag/types/query.py
from __future__ import annotations

from typing import Any, Literal, NotRequired, TypedDict

from .documents import MultimodalElementRef

SearchMode = Literal["semantic", "keyword", "hybrid"]


class QueryBody(TypedDict):
    query: str
    top_k: NotRequired[int]
    filters: NotRequired[dict[str, Any]]
    min_score: NotRequired[float]
    search_mode: NotRequired[SearchMode]
    rerank: NotRequired[bool]
    multimodal: NotRequired[bool]


class QueryResult(TypedDict):
    id: str
    text: str
    score: float
    document_id: str | None
    document_filename: str | None
    chunk_index: int | None
    page_no: int | None
    char_start: int | None
    char_end: int | None
    multimodal_elements: list[MultimodalElementRef]
    metadata: dict[str, Any]


class QueryTimings(TypedDict):
    embed_ms: float
    search_ms: float
    rerank_ms: float
    cache_ms: float
    total_ms: float
    cache_hit: bool


class QueryResponse(TypedDict):
    results: list[QueryResult]
    query: str
    collection: str
    total: int
    timings: NotRequired[QueryTimings | None]


class MultiQueryBody(TypedDict):
    query: str
    collections: list[str]
    top_k: NotRequired[int]
    filters: NotRequired[dict[str, Any]]
    min_score: NotRequired[float]
    search_mode: NotRequired[SearchMode]
    rerank: NotRequired[bool]
    multimodal: NotRequired[bool]


class MultiQueryResult(TypedDict):
    id: str
    text: str
    score: float
    document_id: str | None
    document_filename: str | None
    chunk_index: int | None
    collection: str
    multimodal_elements: list[MultimodalElementRef]
    metadata: dict[str, Any]


class MultiQueryResponse(TypedDict):
    results: list[MultiQueryResult]
    query: str
    collections: list[str]
    total: int


class BatchQueryItem(TypedDict):
    collection: str
    query: str
    top_k: NotRequired[int]
    filters: NotRequired[dict[str, Any]]
    min_score: NotRequired[float]
    search_mode: NotRequired[SearchMode]
    rerank: NotRequired[bool]
    multimodal: NotRequired[bool]


class BatchQueryBody(TypedDict):
    queries: list[BatchQueryItem]


class BatchQueryResultItem(TypedDict):
    results: list[QueryResult]
    query: str
    collection: str
    total: int


class BatchQueryResponse(TypedDict):
    results: list[BatchQueryResultItem]
